feat: migrate store to sqlite

shenlong-tanwen opened the pull request.

shenlong-tanwen committed:
add store entity and migration

shenlong-tanwen committed:
make store service take both isar and drift repos

shenlong-tanwen committed:
migrate and switch store on beta timeline state change

shenlong-tanwen committed:
chore: make drift variables final

shenlong-tanwen committed:
dispose old store before switching repos

shenlong-tanwen commented on mobile/lib/domain/services/store.service.dart:
```diff
@@ -24,13 +34,43 @@ class StoreService {
 
-  static Future<StoreService> create({required IsarStoreRepository storeRepository}) async {
-    final instance = StoreService._(storeRepository: storeRepository);
+  static Future<void> switchRepo(bool isBetaTimeline) async {
+    final current = I;
+    if (current._isBetaTimeline == isBetaTimeline) {
+      return;
+    }
+
+    current.dispose();
+    _instance = await create(
+      storeRepository: current._isarStoreRepository,
+      driftStoreRepository: current._driftStoreRepository,
+      isBetaTimeline: isBetaTimeline,
+    );
+  }
```
Since we don't close and re-open the app, we need this method temporarily to switch the implementations used during runtime.

shenlong-tanwen replied:
This has been removed now and the user is requested to close and re-open the app on migration

shenlong-tanwen commented on mobile/lib/infrastructure/repositories/store.repository.dart:
```diff
@@ -1,23 +1,39 @@
 
-class IsarStoreRepository extends IsarDatabaseRepository {
+// Temporary interface until Isar is removed to make the service work with both Isar and Sqlite
+abstract class IStoreRepository {
+  Future<bool> deleteAll();
+  Stream<StoreDto<Object>> watchAll();
+  Future<void> delete<T>(StoreKey<T> key);
+  Future<bool> insert<T>(StoreKey<T> key, T value);
+  Future<T?> tryGet<T>(StoreKey<T> key);
+  Future<bool> update<T>(StoreKey<T> key, T value);
+  Stream<T?> watch<T>(StoreKey<T> key);
+  Future<List<StoreDto<Object>>> getAll();
+}
```
Same reason this exists as well. We need a common interface to access the store to switch between them during runtime

shenlong-tanwen commented on mobile/lib/widgets/settings/beta_timeline_list_tile.dart:
```diff
@@ -89,7 +89,6 @@ class _BetaTimelineListTileState extends ConsumerState<BetaTimelineListTile> wit
                   ElevatedButton(
                     onPressed: () async {
                       Navigator.of(context).pop();
-                      await ref.read(appSettingsServiceProvider).setSetting(AppSettingsEnum.betaTimeline, value);
```
This is now updated in Change Experience page after the migration is successful

shenlong-tanwen committed:
use store to update values for beta timeline

shenlong-tanwen committed:
change log service to use the proper store

alextran1502 commented:
How do you handle migrating to the new store database when the user is already on the beta timeline?

shenlong-tanwen committed:
migrate store when beta already enabled

shenlong-tanwen commented on mobile/lib/pages/common/change_experience.page.dart:
```diff
@@ -37,46 +39,58 @@ class _ChangeExperiencePageState extends ConsumerState<ChangeExperiencePage> {
   }
 
   Future<void> _handleMigration() async {
```
The code already has a single try-catch block that wraps the entire logic. There are no nested try-catches within this method. On a related note, we do not handle the error gracefully, but the new changes will switch timeline only when the migration is successful whereas the previous one did even when the migration failed.

shenlong-tanwen committed:
use isar repository to check beta timeline in store service

shenlong-tanwen committed:
remove unused update method from store repo

shenlong-tanwen commented on mobile/lib/infrastructure/repositories/store.repository.dart:
```diff
@@ -88,8 +109,99 @@ class IsarStoreRepository extends IsarDatabaseRepository {
+
+  @override
+  Future<bool> update<T>(StoreKey<T> key, T value) async {
+    await _db.storeEntity.update().write(await _fromValue(key, value));
```
lol good catch, but since we never use this method, I've never faced an issue during my testing. Removed the method now

shenlong-tanwen commented:
> How do you handle migrating to the new store database when the user is already on the beta timeline?

Ugh, was reminding myself to add this while working on the changes but forgot it before pushing out the PR. Added it now.

shenlong-tanwen committed:
dispose after create

shenlong-tanwen committed:
change watchAll signature in store repo

shenlong-tanwen commented on mobile/lib/infrastructure/repositories/store.repository.dart:
```diff
@@ -88,8 +100,93 @@ class IsarStoreRepository extends IsarDatabaseRepository {
+      const (String) => (null, value as String),
+      const (bool) => ((value as bool) ? 1 : 0, null),
+      const (DateTime) => ((value as DateTime).millisecondsSinceEpoch, null),
+      const (UserDto) => (null, (await DriftUserRepository(_db).upsert(value as UserDto)).id),
```
Not just odd, it is quite absurd to do this TBH. This is a remanent from the old times that never got refactored. Ideally, we should only store the user id and let the caller take care of fetching and storing the user to the DB. I'll address this in the Auth user PR which affects everything user

shenlong-tanwen replied:
> should we cleanse it with a proper implementation?

For sure, I'll take care of this after the background work. Doing it in this PR means we need to make sure that user is properly stored in the DB before being accessed elsewhere. But doing it in the Auth user PR is easy because the sync stream takes care of adding the user to the table

shenlong-tanwen committed:
fix test

alextran1502 commented on mobile/lib/utils/bootstrap.dart:
```diff
@@ -45,14 +53,23 @@ abstract final class Bootstrap {
+      driftStoreRepository: DriftStoreRepository(drift),
+    );
+
+    final IStoreRepository storeRepo = Store.get(StoreKey.betaTimeline, false)
+        ? DriftStoreRepository(drift)
+        : IsarStoreRepository(db);
```
Sounds good

shenlong-tanwen replied:
This has been migrated to fetch the beta timeline value from Isar instead. Isar is currently expected to have the most updated value for the beta timeline key. The PR also updates the value in both the DBs during migration so it'll always have the proper value even after this PR

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into feat/sqlite-store

alextran1502 commented:
I just ran the first round of testing, found some quirks

When migrating from the old timeline to the beta timeline

- There is an issue with `currentUserProvider` after migration that seems to make it null and kick the users to the login screen

When migrating from the beta timeline back to the old timeline, some errors get triggered as well. See below

```
flutter:
flutter: [SEVERE] [2025-08-20 17:32:54.994893] [ImmichErrorLogger] FlutterError - Catch all
flutter: Error: ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞════════════════════════
flutter: The following StoreKeyNotFoundException was thrown building
flutter: UserCircleAvatar(state: _ConsumerState#fca37):
flutter: Key - <serverEndpoint> not available in Store
flutter:
flutter: The relevant error-causing widget was:
flutter:   UserCircleAvatar
flutter:   UserCircleAvatar:file:///Users/alex/Documents/projects/immich/mobile/lib/widgets/common/immich_app_bar.dart:56:26
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #0      StoreService.get (package:immich_mobile/domain/services/store.service.dart:106:7)
flutter: #1      UserCircleAvatar.build (package:immich_mobile/widgets/common/user_circle_avatar.dart:25:18)
flutter: #2      _ConsumerState.build (package:flutter_riverpod/src/consumer.dart:476:19)
flutter: #3      StatefulElement.build (package:flutter/src/widgets/framework.dart:5823:27)
flutter: #4      ConsumerStatefulElement.build (package:flutter_riverpod/src/consumer.dart:539:20)
flutter: #5      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5715:15)
flutter: #6      StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5874:11)
flutter: #7      Element.rebuild (package:flutter/src/widgets/framework.dart:5427:7)
flutter: #8      StatefulElement.update (package:flutter/src/widgets/framework.dart:5899:5)
flutter: #9      Element.updateChild (package:flutter/src/widgets/framework.dart:3982:15)
flutter: #10     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:7015:14)
flutter: #11     Element.updateChild (package:flutter/src/widgets/framework.dart:3982:15)
```


It is entirely suitable to ask the users to re-login after migration if that makes our lives easier, especially for when migration from beta timeline back to the old one.

shenlong-tanwen committed:
rename init isar to initDB

shenlong-tanwen committed:
request user to close and reopen on beta migration

shenlong-tanwen commented:
> It is entirely suitable to ask the users to re-login after migration if that makes our lives easier, especially for when migration from beta timeline back to the old one.

This would make things a lot easier and straightforward. I've updated the migration to now ask the users to close and re-open the app after the migration. On the next re-open the correct repo is selected and used for the Store and Log services. 

I've left the two way migration in as we currently store the Server URL, Access Token and Current User, all inside the Store. If we don't keep this updated, there might be cases like the error you've faced where a user might be dropped back to the login screen after the migration back to the old timeline

The change experience page is also updated to show the error message when the migration fails. Previously, it gets stuck at an infinite loading indicator without any visual feedback

shenlong-tanwen committed:
fix tests

shenlong-tanwen committed:
handle empty version in migration

alextran1502 committed:
Merge branch 'main' into feat/sqlite-store

shenlong-tanwen committed:
wait for cache to be populated after migration
